### ISSUE TITLE
fix(model,vulkan): qualify bounded F32-row Qwen sequence prefill parity

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12695]
# fix(model,vulkan): qualify bounded F32-row Qwen sequence prefill parity

```routing
lane: model
paths: ["internal/model/qwen35_hal.go", "internal/model/qwen35_sequence_embedding_rows.go", "internal/model/qwen35_sequence_embedding_rows_test.go", "internal/model/qwen35_sequence_prefill_test.go", "internal/compute/shaders/qwen35_causal_attention_panel.comp", "internal/compute/vulkan_qwen35_attention_order_test.go"]
expected_steps: 8
```

**Work Unit:** leaf  
**Priority:** P0

## Parent context

Parent: #12177. This issue owns the bounded model-admission and causal-attention parity prerequisites for exercising sequence prefill through an actual checkpoint. The experimental Q4_K cooperative candidate remains outside this scope and disabled.

## Current state

At public base `816506b53c263edd44e4b2a151ac80cac3b44031`, the Qwen sequence-prefill admission checked the entire 248,320 × 5,120 F32 embedding table against the backend's single-buffer cap. Its 5,085,593,600 bytes exceed Vulkan's maximum representable storage-buffer range, so the runtime safely declined sequence prefill even when a small prompt-row panel fit.

After bounded row admission first reached hardware, a source-bound model run exposed a separate numerical defect: sequence causal attention changed the online-softmax operation order from the existing token path. Generated token IDs matched, but full logits exceeded the parity gate. The repaired shader preserves decode's per-key QK reduction and sequential online-softmax order.

Public commit `bf0078366e64d48050ba64d4179a5b7419072072` ships the six-file row-admission and attention-order repair. Public commit `762626f884f428dd99cae8cd6e9ef72e801f7b92` is the current containing commit after the adjacent cooperative Q8 correctness repair.

## Why this is next

The full embedding-table allocation blocked the actual checkpoint before native sequence execution. Once that barrier was removed, attention-order parity became the correctness gate for any whole-model prefill measurement. Both defects must remain fixed before #12177 can evaluate an experimental Q4_K route.

## Core through-line

For an F32 embedding table that exceeds the device cap, validate the prompt-row panel against the same cap, gather ordered and scaled rows on the host, upload only that bounded panel, and invoke the existing native sequence path. Resolve and cap-check the output-head representation before dispatch. In sequence causal attention, preserve the token path's arithmetic order. Return explicit fallback when the panel, output head, row capability, or native route is unavailable.

## Working spine

Prompt token IDs enter the existing session prefill API, become a bounded embedding-row tensor, reach native sequence execution, pass through token-order causal attention, and produce logits plus route attribution. No new public interface is introduced.

## Gold-plating boundary

No new quantization format, full-table sharding, cache policy, speculative algorithm, or device-specific tuning. Reuse the existing sequence-row capability and model embedding oracle. Repair only the existing causal-attention shader's operation order. Do not alter or enable the cooperative Q4_K selector.

## Value

- Centrality: Core.
- P1: advanced — removes the deterministic allocation barrier and restores native sequence correctness.
- P2: advanced — device upload scales with prompt length rather than vocabulary size.
- P3: advanced — embedding rows, logits, greedy IDs, alternate prompts, and append continuation are checked against existing oracles.
- P4: preserved — confined to public `internal/model` and `internal/compute`; SDK and private serving interfaces are unchanged.

## Witness

Run in the public repository:

```bash
go test -count=1 -run '^TestQwen35SequencePrefillF32EmbeddingRows' ./internal/model
go test -tags vulkan -count=1 -run '^TestVulkanQwen35AttentionPanelMatchesTokenOrder$' ./internal/compute
```

The independent row regression failed before the production call-site change because the former route made zero batched sequence calls. The final fixture uses a genuinely packed in-cap output head, records every upload, and covers ordered, repeated, scaled, invalid-ID, over-budget, Q2, and cleanup behavior.

The independent required-device attention regression compares growing-KV token execution with the sequence panel for prefix lengths 0, 17, 129, and 257. The final production shader is bitwise exact for all four; the rebuilt base shader fails every exact gate. Source-bound 2026-09-09 whole-model runs then recorded exact equality for five compared vectors and greedy IDs, an alternate 32-token prompt, and a 16+16 append. They required native routing, no fallback, and zero experimental Q4_K dispatches. The final six-pair run binds public commit `762626f884f428dd99cae8cd6e9ef72e801f7b92`, its test-only witness overlay, and all shipping shaders; all correctness gates passed. Its timing remains a dated single-host snapshot rather than a permanent capability claim.

### File:Line Seam

- `internal/model/qwen35_hal.go:934` (`tryQwen35SequencePrefill`, route and allocation admission).
- `internal/model/qwen35_sequence_embedding_rows.go:13` (`qwen35EmbeddingRows`, ordered host gather).
- `internal/compute/shaders/qwen35_causal_attention_panel.comp:31` (`main`, causal attention arithmetic order).
- `internal/model/qwen35_sequence_embedding_rows_test.go:66` (bounded-row regression).
- `internal/compute/vulkan_qwen35_attention_order_test.go:16` (required-device token-order regression).

### Blast Radius & Affected Lanes

Only public `internal/model` and `internal/compute` change. Model loaders, SDKs, private serving policy, and unrelated lanes remain unchanged.

### Quarantined Fallback Mechanism

The existing per-token path remains the fallback for a missing row capability, over-budget panel, over-budget resolved output head, or unsupported sequence route. Whole-table safety remains intact. The experimental Q4_K candidate remains opt-in and does not contribute to this issue's correctness witness.

## Done condition

The bounded F32 row path reaches native sequence execution, stays within the buffer cap, matches token-order causal attention, preserves continuation, and keeps unsupported requests on explicit fallback.

### Scoped Acceptance Criteria

- [x] [SW-VERIFIED] Bounded F32 rows preserve first, last, repeated, and scaled values without uploading the full input table.
- [x] [SW-VERIFIED] Panel and output-head cap checks occur before allocation; invalid token IDs fail explicitly; cleanup is bounded.
- [x] [SW-VERIFIED] Full-table, Q2-row, absent-capability, and over-budget fallbacks remain covered.
- [x] [SW-VERIFIED] Sequence attention is bitwise exact against token-order Vulkan execution for the qualified prefixes.
- [x] [HW-WITNESSED] Source-bound device runs pass logits, generated-token, alternate-prompt, and 16+16 continuation comparisons with native routing and no fallback.
- [x] [SW-VERIFIED] Fresh-checkout commit `762626f884f428dd99cae8cd6e9ef72e801f7b92` rebuilds from source; untagged compute and required-device Q8 plus attention regressions pass with zero required-device skips.
- [x] [HW-WITNESSED] The final exact-commit target-device run passes all six P128 pairs plus alternate-prompt and 16+16 append correctness with native routing and no fallback.
- [x] [HW-WITNESSED] The pinned upgrade runs the exact source-bound V10 artifact set; smoke, boundary, cancellation, recovery, and idle readback gates pass.

The earlier failed model run is retained as diagnostic history: bounded rows reached native execution with no fallback and matching greedy IDs, while logits exceeded the parity ceiling. It supports no performance or default-promotion claim. Repeated timing remains snapshot-scoped; no throughput floor or speedup is asserted by this issue.

## Closure evidence

Public commits `bf0078366e64d48050ba64d4179a5b7419072072` and `762626f884f428dd99cae8cd6e9ef72e801f7b92` contain the qualified model, attention, and Q8 prerequisites. The final exact-source Halo build passed six paired P128 comparisons, the alternate P32 prompt, and 16+16 continuation with native routing, no fallback, and exact numerical gates. The pinned deployment retained that artifact identity and passed five smoke requests, a separate 283-input-token boundary request, deliberate client cancellation with server acknowledgement and quiescent lease release, recovery, and a healthy idle readback. The experimental Q4_K candidate remained absent and is still tracked separately by #12177.

## Likely files

- `internal/model/qwen35_hal.go`
- `internal/model/qwen35_sequence_embedding_rows.go`
- `internal/model/qwen35_sequence_embedding_rows_test.go`
- `internal/model/qwen35_sequence_prefill_test.go`
- `internal/compute/shaders/qwen35_causal_attention_panel.comp`
- `internal/compute/vulkan_qwen35_attention_order_test.go`

## Lane

Public `model` and `compute`; two packages, eight bounded steps.

## Acceptance gate

Both independent fail-before/pass-after regressions pass, fallback and Q2 behavior remain green, and whole-model prompt plus append comparisons are exact. A skipped GPU test cannot satisfy the physical criterion. Closure also requires an exact public-commit rebuild, native regression, and deployment readback.

## Closure binding

The signed implementation commits reference this issue and carry separated implementation/test provenance. Close only after the exact public containing commit passes the final physical checks.

**Comment by anthony-chaudhary** [comment]
Closed by public commits `bf0078366e64d48050ba64d4179a5b7419072072` and `762626f884f428dd99cae8cd6e9ef72e801f7b92`.

The exact-source Halo build passed all six P128 pairs plus alternate-prompt and 16+16 continuation gates with native routing and no fallback. The pinned deployment then passed smoke, a separate 283-input-token boundary request, cancellation/quiescence, recovery, and healthy idle readback. The experimental Q4_K candidate stayed disabled and remains tracked by #12177.
